File: Structures/LinkedList/Go/armin/linked-list.go
package linkedList

import (
	"fmt"
)

type Node struct {
	key  interface{}
	next *Node
	prev *Node
}

type LinkedList struct {
	size int
	head *Node
	tail *Node
}

func CreateLinkedList() *LinkedList {
	list := LinkedList{
		size: 0,
		head: nil,
		tail: nil,
	}
	return &list
}

func (list *LinkedList) AddFront(key interface{}) {
	if list.size == 0 {
		list.firstInitialize(key)
	} else if list.size == 1 {
		node := &Node{
			key:  key,
			prev: list.head,
			next: nil,
		}
		list.head.next = node
		list.tail = node
	} else {
		node := &Node{
			key:  key,
			prev: list.tail,
			next: nil,
		}
		list.tail.next = node
		list.tail = node
	}
	list.size = list.size + 1
}

func (list *LinkedList) AddBack(key interface{}) {
	if list.size == 0 {
		list.firstInitialize(key)
	} else if list.size == 1 {
		node := &Node{
			key:  key,
			prev: nil,
			next: list.head,
		}
		list.head.prev = node
		list.tail = list.head
		list.head = node
	} else {
		node := &Node{
			key:  key,
			prev: nil,
			next: list.head,
		}
		list.head.prev = node
		list.head = node
	}
	list.size = list.size + 1
}

// adding first element to linked list
func (list *LinkedList) firstInitialize(key interface{}) {
	node := &Node{
		key:  key,
		prev: nil,
		next: nil,
	}
	list.head = node
}

func (list *LinkedList) Delete(index int) *Node {
	node := list.Search(index)
	if node == nil {
		return nil
	} else {
		list.size = list.size - 1
		if node.next == nil { //node is tail of list
			list.tail = node.prev
			node.prev.next = nil
		} else if node.prev == nil { //nod is head of list
			list.head = node.next
			node.next.prev = nil
		} else {
			node.next.prev = node.prev
			node.prev.next = node.next
		}
		return node
	}
}

func (list *LinkedList) Display() {
	rooter := list.head
	fmt.Print("HEAD OF LIST : ")
	for {
		fmt.Println(rooter.key)
		rooter = rooter.next
		if rooter.next == nil {
			fmt.Print("TAIL OF LIST : ")
			fmt.Println(rooter.key)
			break
		}
	}

}

func (list *LinkedList) Search(index int) *Node {
	if index < 0 || index > list.size {
		fmt.Println("INVALID INDEX !")
		return nil
	} else {
		node := list.head
		for i := 0; i < index-1; i++ {
			node = node.next
		}
		return node
	}
}

func main() {
	link := CreateLinkedList()
	link.AddFront("head")
	link.AddFront(30)
	link.AddFront(40)
	link.AddFront(50)
	link.AddBack("new head")
	link.AddBack("new new head")
	link.AddFront("tail")
	link.Display()
	fmt.Println("######################################################3")
	link.Delete(10)
	link.Display()
}

File: Structures/HashTable/CPP/LinkedList.cpp
#include <iostream>
using namespace std;

class Node{
    public:
    int key;
    Node* next;
    Node* prev;
    Node(int key){
        this->key=key;
        next=NULL;
        prev=NULL;
    }
};


class LinkedList{
    private:
        Node* head;
        Node* tail;
        Node* createNode(int key){
            Node* newNode=new Node(key);
            return newNode;
        }
    public:
        LinkedList(){
            head=NULL;
            tail=NULL;
        }
        void insert_key(int key){
            Node* newNode=createNode(key);
            if(head==NULL){
                head=tail=newNode;
                return;
            }
            newNode->next=head;
            head->prev=newNode;
            head=newNode;
            return;
        };
        Node* delete_key(int key){
            Node* nodeToDel=search_key(key);
            if(nodeToDel==head){
                nodeToDel->next->prev=NULL;
                head=nodeToDel->next;
                //delete(nodeToDel);
            }else if(nodeToDel==tail){
                nodeToDel->prev->next=NULL;
                tail=nodeToDel->prev;
                //delete(nodeToDel);
            }else{
                nodeToDel->prev->next=nodeToDel->next;
                nodeToDel->next->prev=nodeToDel->prev;
                //delete(nodeToDel);
            }
            return nodeToDel;
        };
        Node* search_key(int key){
            if(head==NULL)return NULL;
            Node* currentNode=head;
            while(currentNode!=NULL){
                if(currentNode->key==key)return currentNode;
                currentNode=currentNode->next;
            }
            return NULL;

        };
        void print_list(){
            Node* current=head;
            while(current!=NULL){
                cout<<current->key<<" ";
            }
        };
};